util/feeding.py

```diff
@@ -1,198 +1,97 @@
+    return source_data
+
+
+def samples_to_mfccs(samples, sample_rate):
+    spectrogram = contrib_audio.audio_spectrogram(samples, window_size=512, stride=320, magnitude_squared=True)
+    mfccs = contrib_audio.mfcc(spectrogram, sample_rate, dct_coefficient_count=Config.n_input)
+    mfccs = tf.reshape(mfccs, [-1, Config.n_input])
+
+    return mfccs, tf.shape(mfccs)[0]
+
+
+def audiofile_to_features(wav_filename):
+    samples = tf.read_file(wav_filename)
+    decoded = contrib_audio.decode_wav(samples, desired_channels=1)
+    features, features_len = samples_to_mfccs(decoded.audio, decoded.sample_rate)
+
+    return features, features_len
+
+
+def entry_to_features(wav_filename, transcript):
+    # https://bugs.python.org/issue32117
+    features, features_len = audiofile_to_features(wav_filename)
+    return features, features_len, tf.SparseTensor(*transcript)
+
+
+def to_sparse_tuple(sequence):
+    r"""Creates a sparse representention of ``sequence``.
+        Returns a tuple with (indices, values, shape)
+    """
+    indices = np.asarray(list(zip([0]*len(sequence), range(len(sequence)))), dtype=np.int64)
+    shape = np.asarray([1, len(sequence)], dtype=np.int64)
+    return indices, sequence, shape
+
+
+def create_dataset(csvs, batch_size, cache_path):
+    df = read_csvs(csvs)
+    df.sort_values(by='wav_filesize', inplace=True)
+
+    num_batches = len(df) // batch_size
+
+    # Convert to character index arrays
+    df['transcript'] = df['transcript'].apply(partial(text_to_char_array, alphabet=Config.alphabet))
+
+    def generate_values():
+        for _, row in df.iterrows():
+            yield row.wav_filename, to_sparse_tuple(row.transcript)
+
+    # Batching a dataset of 2D SparseTensors creates 3D batches, which fail
+    # when passed to tf.nn.ctc_loss, so we reshape them to remove the extra
+    # dimension here.
+    def sparse_reshape(sparse):
+        shape = sparse.dense_shape
+        return tf.sparse.reshape(sparse, [shape[0], shape[2]])
+
+    def batch_fn(features, features_len, transcripts):
+        features = tf.data.Dataset.zip((features, features_len))
+        features = features.padded_batch(batch_size,
+                                         padded_shapes=([None, Config.n_input], []))
+        transcripts = transcripts.batch(batch_size).map(sparse_reshape)
+        return tf.data.Dataset.zip((features, transcripts))
+
+    num_gpus = len(Config.available_devices)
+
+    dataset = (tf.data.Dataset.from_generator(generate_values,
+                                              output_types=(tf.string, (tf.int64, tf.int32, tf.int64)))
+                              .map(entry_to_features, num_parallel_calls=tf.data.experimental.AUTOTUNE)
+                              .cache(cache_path)
+                              .window(batch_size, drop_remainder=True).flat_map(batch_fn)
+                              .prefetch(num_gpus)
+                              .repeat())
 
+    return dataset, num_batches
```

Comment:
I think it's time to rename this file, as it is not "feeding" anymore.

Reply:
@tilmankamp @kdavis-mozilla any suggestions? It's still handling feeding data into the network in my mind, I don't know what else to call it or in what direction it needs clarification.